File: sf_core/tests/e2e/authentication/native_okta.rs
use crate::common::snowflake_test_client::SnowflakeTestClient;

// Native Okta E2E tests require VPN access to preprod Snowflake account.
// Skipped in GH Actions (no VPN), run on Jenkins or locally with:
//   PARAMETER_PATH=parameters.json cargo test -- --ignored vpn_

#[test]
#[ignore = "Requires VPN to access preprod account (run on Jenkins)"]
fn vpn_should_authenticate_using_native_okta() {
    // Given Okta authentication is configured with valid credentials
    let client = SnowflakeTestClient::with_default_params();
    let okta_url = client
        .parameters
        .okta_url
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_URL must be set for Okta E2E tests");
    let okta_user = client
        .parameters
        .okta_user
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_USER must be set for Okta E2E tests");
    let okta_password = client
        .parameters
        .okta_password
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_PASSWORD must be set for Okta E2E tests");
    let okta_account = client
        .parameters
        .okta_account
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_ACCOUNT must be set for Okta E2E tests");
    let okta_host = client
        .parameters
        .okta_host
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_HOST must be set for Okta E2E tests");

    // Use the Okta-enabled Snowflake account with its specific settings
    client.set_connection_option("account", &okta_account);
    client.set_connection_option("host", &okta_host);
    client.set_connection_option("authenticator", &okta_url);
    client.set_connection_option("user", &okta_user);
    client.set_connection_option("password", &okta_password);
    // Use PUBLIC role to avoid role mismatch errors
    client.set_connection_option("role", "PUBLIC");

    // When Trying to Connect
    let result = client.connect();

    // Then Login is successful and simple query can be executed
    client.verify_simple_query(result);
}

#[test]
#[ignore = "Requires VPN to access preprod account (run on Jenkins)"]
fn vpn_should_fail_native_okta_authentication_with_wrong_credentials() {
    // Given Okta authentication is configured with wrong password
    let client = SnowflakeTestClient::with_default_params();
    let okta_url = client
        .parameters
        .okta_url
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_URL must be set for Okta E2E tests");
    let okta_user = client
        .parameters
        .okta_user
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_USER must be set for Okta E2E tests");
    let okta_account = client
        .parameters
        .okta_account
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_ACCOUNT must be set for Okta E2E tests");
    let okta_host = client
        .parameters
        .okta_host
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_HOST must be set for Okta E2E tests");

    // Use the Okta-enabled Snowflake account with its specific settings
    client.set_connection_option("account", &okta_account);
    client.set_connection_option("host", &okta_host);
    client.set_connection_option("authenticator", &okta_url);
    client.set_connection_option("user", &okta_user);
    client.set_connection_option("password", "wrong_password_12345");
    client.set_connection_option("role", "PUBLIC");

    // When Trying to Connect
    let result = client.connect();

    // Then Connection fails with authentication error
    client.assert_login_error(result);
}

#[test]
#[ignore = "Requires VPN to access preprod account (run on Jenkins)"]
fn vpn_should_fail_native_okta_authentication_with_wrong_okta_url() {
    // Given Okta authentication is configured with invalid okta url
    let client = SnowflakeTestClient::with_default_params();
    let okta_user = client
        .parameters
        .okta_user
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_USER must be set for Okta E2E tests");
    let okta_password = client
        .parameters
        .okta_password
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_PASSWORD must be set for Okta E2E tests");
    let okta_account = client
        .parameters
        .okta_account
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_ACCOUNT must be set for Okta E2E tests");
    let okta_host = client
        .parameters
        .okta_host
        .clone()
        .expect("SNOWFLAKE_TEST_OKTA_HOST must be set for Okta E2E tests");

    // Use the Okta-enabled Snowflake account with invalid Okta URL
    client.set_connection_option("account", &okta_account);
    client.set_connection_option("host", &okta_host);
    client.set_connection_option("authenticator", "https://invalid.okta.com");
    client.set_connection_option("user", &okta_user);
    client.set_connection_option("password", &okta_password);
    client.set_connection_option("role", "PUBLIC");

    // When Trying to Connect
    let result = client.connect();

    // Then Connection fails with authentication error
    client.assert_login_error(result);
}

File: odbc_tests/tests/e2e/authentication/native_okta.cpp
#include <picojson.h>
#include <sql.h>
#include <sqlext.h>
#include <sqltypes.h>

#include <sstream>

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_all.hpp>

#include "HandleWrapper.hpp"
#include "compatibility.hpp"
#include "get_diag_rec.hpp"
#include "macros.hpp"
#include "require.hpp"
#include "test_setup.hpp"

using namespace Catch::Matchers;

// =============================================================================
// Helpers
// =============================================================================

std::string get_okta_connection_string() {
  auto params = get_test_parameters("testconnection");
  std::stringstream ss;
  configure_driver_string(ss);
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_HOST", "SERVER");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_ACCOUNT", "ACCOUNT");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_USER", "UID");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_PASSWORD", "PWD");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_URL", "AUTHENTICATOR");
  ss << "ROLE=PUBLIC;";
  return ss.str();
}

EnvironmentHandleWrapper setup_okta_environment() {
  EnvironmentHandleWrapper env;
  SQLRETURN ret = SQLSetEnvAttr(env.getHandle(), SQL_ATTR_ODBC_VERSION, (SQLPOINTER)SQL_OV_ODBC3, 0);
  CHECK_ODBC(ret, env);
  return env;
}

ConnectionHandleWrapper get_okta_connection_handle(EnvironmentHandleWrapper& env) {
  return env.createConnectionHandle();
}

void attempt_okta_connection(ConnectionHandleWrapper& dbc, const std::string& connection_string) {
  SQLRETURN ret = SQLDriverConnect(dbc.getHandle(), NULL, (SQLCHAR*)connection_string.c_str(), SQL_NTS, NULL, 0, NULL,
                                   SQL_DRIVER_NOPROMPT);
  CHECK_ODBC(ret, dbc);
}

void verify_okta_simple_query_execution(ConnectionHandleWrapper& dbc) {
  StatementHandleWrapper stmt = dbc.createStatementHandle();
  SQLRETURN ret = SQLExecDirect(stmt.getHandle(), (SQLCHAR*)"SELECT 1", SQL_NTS);
  CHECK_ODBC(ret, stmt);

  ret = SQLFetch(stmt.getHandle());
  CHECK_ODBC(ret, stmt);

  SQLINTEGER result = 0;
  ret = SQLGetData(stmt.getHandle(), 1, SQL_C_LONG, &result, sizeof(result), NULL);
  CHECK_ODBC(ret, stmt);
  REQUIRE(result == 1);
}

// =============================================================================
// E2E Tests
// =============================================================================

TEST_CASE("should authenticate using native okta", "[native_okta]") {
  REQUIRE_VPN("Native Okta E2E tests need access to preprod Snowflake account");
  std::string connection_string = get_okta_connection_string();

  // Given Okta authentication is configured with valid credentials
  auto env = setup_okta_environment();
  auto dbc = get_okta_connection_handle(env);

  // When Trying to Connect
  attempt_okta_connection(dbc, connection_string);

  // Then Login is successful and simple query can be executed
  verify_okta_simple_query_execution(dbc);

  SQLDisconnect(dbc.getHandle());
}

TEST_CASE("should fail native okta authentication with wrong credentials", "[native_okta]") {
  REQUIRE_VPN("Native Okta E2E tests need access to preprod Snowflake account");

  // Given Okta authentication is configured with wrong password
  auto params = get_test_parameters("testconnection");
  std::stringstream ss;
  configure_driver_string(ss);
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_HOST", "SERVER");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_ACCOUNT", "ACCOUNT");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_USER", "UID");
  ss << "PWD=wrong_password_12345;";
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_URL", "AUTHENTICATOR");
  ss << "ROLE=PUBLIC;";
  std::string connection_string = ss.str();

  // When Trying to Connect
  auto records = require_connection_failed(connection_string);

  // Then Connection fails with authentication error
  REQUIRE(records.size() >= 1);
  CHECK(records[0].sqlState == "28000");
  CHECK_THAT(records[0].messageText, ContainsSubstring("rejected credentials"));
}

TEST_CASE("should fail native okta authentication with wrong okta url", "[native_okta]") {
  REQUIRE_VPN("Native Okta E2E tests need access to preprod Snowflake account");

  // Given Okta authentication is configured with invalid okta url
  auto params = get_test_parameters("testconnection");
  std::stringstream ss;
  configure_driver_string(ss);
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_HOST", "SERVER");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_ACCOUNT", "ACCOUNT");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_USER", "UID");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_OKTA_PASSWORD", "PWD");
  ss << "AUTHENTICATOR=https://invalid.okta.com;";
  ss << "ROLE=PUBLIC;";
  std::string connection_string = ss.str();

  // When Trying to Connect
  auto records = require_connection_failed(connection_string);

  // Then Connection fails with authentication error
  REQUIRE(records.size() >= 1);
  CHECK(records[0].sqlState == "28000");
  CHECK_THAT(records[0].messageText,
             ContainsSubstring("does not match configured Okta URL") || ContainsSubstring("Native Okta SSO failed"));
}
